fraud/service.go

```diff
@@ -88,57 +136,79 @@ func (f *service) processIncoming(
 		}
 		return pubsub.ValidationReject
 	}
+	// check the fraud proof locally and ignore if it has been already stored locally.
+	if f.verifyLocal(ctx, proofType, hex.EncodeToString(proof.HeaderHash()), msg.Data) {
+		return pubsub.ValidationIgnore
+	}
+
+	msg.ValidatorData = proof
 
-	newCtx, cancel := context.WithTimeout(ctx, fetchHeaderTimeout)
-	extHeader, err := f.getter(newCtx, proof.Height())
-	defer cancel()
+	// fetch extended header in order to verify the fraud proof.
+	extHeader, err := f.getter(ctx, proof.Height())
 	if err != nil {
-		// Timeout means there is a problem with the network.
-		// As we cannot prove or discard Fraud Proof, user must restart the node.
-		if errors.Is(err, context.DeadlineExceeded) {
-			log.Errorw("failed to fetch header. Timeout reached.")
-			// TODO(@vgonkivs): add handling for this case. As we are not able to verify fraud proof.
-		}
 		log.Errorw("failed to fetch header to verify a fraud proof",
 			"err", err, "proofType", proof.Type(), "height", proof.Height())
 		return pubsub.ValidationIgnore
 	}
+	// validate the fraud proof.
+	// Peer will be added to black list if the validation fails.
 	err = proof.Validate(extHeader)
 	if err != nil {
 		log.Errorw("proof validation err: ",
 			"err", err, "proofType", proof.Type(), "height", proof.Height())
 		f.pubsub.BlacklistPeer(from)
 		return pubsub.ValidationReject
 	}
-	log.Warnw("received fraud proof", "proofType", proof.Type(),
-		"height", proof.Height(),
-		"hash", hex.EncodeToString(extHeader.DAH.Hash()),
-		"from", from.String(),
-	)
-	msg.ValidatorData = proof
+
+	// add the fraud proof to storage.
+	err = f.put(ctx, proof.Type(), hex.EncodeToString(proof.HeaderHash()), msg.Data)
+	if err != nil {
+		log.Errorw("failed to store fraud proof", "err", err)
+	}
+
+	return pubsub.ValidationAccept
+}
+
+func (f *ProofService) Get(ctx context.Context, proofType ProofType) ([]Proof, error) {
```

Comment:
I'd godoc this

Reply:
[9577e1d](https://github.com/celestiaorg/celestia-node/pull/917/commits/9577e1dc87fdaea65ba3670a2f94e7dc2dac4d4f#diff-3cb5c912632890410d9c258b1f26bdf0e5b59872cfb519613412cdb05935fb18R44)